File: generate-icons.js
const fs = require('fs');
const path = require('path');
const zlib = require('zlib');

// CRC32 lookup table
const crcTable = (() => {
  const table = new Uint32Array(256);
  for (let n = 0; n < 256; n++) {
    let c = n;
    for (let k = 0; k < 8; k++) {
      c = (c & 1) ? (0xEDB88320 ^ (c >>> 1)) : (c >>> 1);
    }
    table[n] = c;
  }
  return table;
})();

function crc32(buf) {
  let crc = 0xFFFFFFFF;
  for (let i = 0; i < buf.length; i++) {
    crc = crcTable[(crc ^ buf[i]) & 0xFF] ^ (crc >>> 8);
  }
  return (crc ^ 0xFFFFFFFF) >>> 0;
}

function createPNG(width, height, pixels) {
  // pixels is a function (x, y) => [r, g, b, a]
  const raw = [];
  for (let y = 0; y < height; y++) {
    raw.push(0); // filter byte
    for (let x = 0; x < width; x++) {
      const [r, g, b, a] = pixels(x, y);
      raw.push(r, g, b, a);
    }
  }
  const rawBuf = Buffer.from(raw);
  const compressed = zlib.deflateSync(rawBuf);

  const chunks = [];

  // PNG signature
  chunks.push(Buffer.from([137, 80, 78, 71, 13, 10, 26, 10]));

  // IHDR
  const ihdr = Buffer.alloc(13);
  ihdr.writeUInt32BE(width, 0);
  ihdr.writeUInt32BE(height, 4);
  ihdr[8] = 8; // bit depth
  ihdr[9] = 6; // color type: RGBA
  ihdr[10] = 0; ihdr[11] = 0; ihdr[12] = 0;

  chunks.push(createChunk('IHDR', ihdr));

  // IDAT
  chunks.push(createChunk('IDAT', compressed));

  // IEND
  chunks.push(createChunk('IEND', Buffer.alloc(0)));

  return Buffer.concat(chunks);
}

function createChunk(type, data) {
  const len = Buffer.alloc(4);
  len.writeUInt32BE(data.length, 0);

  const typeAndData = Buffer.concat([Buffer.from(type, 'ascii'), data]);

  const crc = Buffer.alloc(4);
  crc.writeUInt32BE(crc32(typeAndData), 0);

  return Buffer.concat([len, typeAndData, crc]);
}

// Icon drawing helpers
function drawCircle(pixels, cx, cy, r, color) {
  return (x, y) => {
    const dx = x - cx;
    const dy = y - cy;
    const dist = Math.sqrt(dx * dx + dy * dy);
    if (dist <= r) {
      // Anti-aliasing
      const edge = Math.max(0, Math.min(1, r - dist + 0.5));
      return color.map(c => Math.round(c * edge));
    }
    return [0, 0, 0, 0];
  };
}

function drawRect(pixels, x1, y1, x2, y2, color) {
  return (x, y) => {
    if (x >= x1 && x <= x2 && y >= y1 && y <= y2) {
      return [...color];
    }
    return pixels ? pixels(x, y) : [0, 0, 0, 0];
  };
}

function drawLine(pixels, x1, y1, x2, y2, color, thickness) {
  return (x, y) => {
    const dist = pointToLineDist(x, y, x1, y1, x2, y2);
    if (dist <= thickness / 2) {
      return [...color];
    }
    return pixels ? pixels(x, y) : [0, 0, 0, 0];
  };
}

function pointToLineDist(px, py, x1, y1, x2, y2) {
  const A = px - x1;
  const B = py - y1;
  const C = x2 - x1;
  const D = y2 - y1;
  const dot = A * C + B * D;
  const lenSq = C * C + D * D;
  let param = lenSq !== 0 ? dot / lenSq : -1;
  let xx, yy;
  if (param < 0) { xx = x1; yy = y1; }
  else if (param > 1) { xx = x2; yy = y2; }
  else { xx = x1 + param * C; yy = y1 + param * D; }
  return Math.sqrt((px - xx) ** 2 + (py - yy) ** 2);
}

function combinePixels(width, height, layers) {
  return (x, y) => {
    for (let i = layers.length - 1; i >= 0; i--) {
      const [r, g, b, a] = layers[i](x, y);
      if (a > 0) return [r, g, b, a];
    }
    return [0, 0, 0, 0];
  };
}

// Create icons
const SIZE = 81;
const GRAY = [138, 147, 132, 255]; // #8a9384
const GREEN = [47, 125, 79, 255]; // #2f7d4f

function createGroupIcon(color) {
  const layers = [
    // Person 1 (left)
    drawCircle(null, 32, 28, 10, color),
    drawRect(null, 20, 45, 44, 70, color),
    // Person 2 (right)
    drawCircle(null, 52, 28, 10, color),
    drawRect(null, 40, 45, 64, 70, color),
  ];
  return combinePixels(SIZE, SIZE, layers);
}

function createCheckinIcon(color) {
  const layers = [
    // Calendar base
    drawRect(null, 15, 25, 65, 65, color),
    // Checkmark
    drawLine(null, 25, 45, 38, 58, [255, 255, 255, 255], 6),
    drawLine(null, 38, 58, 58, 30, [255, 255, 255, 255], 6),
  ];
  return combinePixels(SIZE, SIZE, layers);
}

function createMeIcon(color) {
  const layers = [
    drawCircle(null, 40, 28, 14, color),
    drawRect(null, 22, 50, 58, 72, color),
  ];
  return combinePixels(SIZE, SIZE, layers);
}

function createReviewIcon(color) {
  const layers = [
    // Bar chart
    drawRect(null, 18, 50, 30, 70, color),
    drawRect(null, 35, 40, 47, 70, color),
    drawRect(null, 52, 30, 64, 70, color),
  ];
  return combinePixels(SIZE, SIZE, layers);
}

// Generate icon files
const icons = [
  { name: 'tab-group', normal: createGroupIcon(GRAY), selected: createGroupIcon(GREEN) },
  { name: 'tab-checkin', normal: createCheckinIcon(GRAY), selected: createCheckinIcon(GREEN) },
  { name: 'tab-me', normal: createMeIcon(GRAY), selected: createMeIcon(GREEN) },
  { name: 'tab-review', normal: createReviewIcon(GRAY), selected: createReviewIcon(GREEN) },
];

const outDir = path.join(__dirname, '..', 'program', 'images', 'tab-icons');
if (!fs.existsSync(outDir)) {
  fs.mkdirSync(outDir, { recursive: true });
}

for (const icon of icons) {
  const normalPath = path.join(outDir, `${icon.name}.png`);
  const selectedPath = path.join(outDir, `${icon.name}-active.png`);

  fs.writeFileSync(normalPath, createPNG(SIZE, SIZE, icon.normal));
  fs.writeFileSync(selectedPath, createPNG(SIZE, SIZE, icon.selected));

  console.log(`Created: ${normalPath}`);
  console.log(`Created: ${selectedPath}`);
}

console.log('\nAll icons created successfully!');

File: generate-icons.py
#!/usr/bin/env python3
"""Generate minimal PNG icons for WeChat Mini Program tab bar."""

import struct
import zlib
import os

def create_png(width, height, pixels_func):
    """Create a PNG file from a pixel function."""
    # PNG signature
    signature = b'\x89PNG\r\n\x1a\n'

    # IHDR chunk
    ihdr_data = struct.pack('>IIBBBBB', width, height, 8, 6, 0, 0, 0)
    ihdr = make_chunk('IHDR', ihdr_data)

    # IDAT chunk - image data
    raw_data = b''
    for y in range(height):
        raw_data += b'\x00'  # filter byte: None
        for x in range(width):
            r, g, b, a = pixels_func(x, y)
            raw_data += struct.pack('BBBB', r, g, b, a)

    compressed = zlib.compress(raw_data)
    idat = make_chunk('IDAT', compressed)

    # IEND chunk
    iend = make_chunk('IEND', b'')

    return signature + ihdr + idat + iend

def make_chunk(chunk_type, data):
    """Create a PNG chunk."""
    chunk = chunk_type.encode('ascii') + data
    length = struct.pack('>I', len(data))
    crc = struct.pack('>I', zlib.crc32(chunk) & 0xFFFFFFFF)
    return length + chunk + crc

def draw_circle(cx, cy, r, color):
    """Return a pixel function that draws a circle."""
    def pixel_func(x, y):
        dx = x - cx
        dy = y - cy
        dist = (dx*dx + dy*dy) ** 0.5
        if dist <= r:
            # Simple anti-aliasing
            alpha = max(0, min(255, int(color[3] * (1 - (dist - r + 0.5) / 1))))
            return (color[0], color[1], color[2], alpha)
        return (0, 0, 0, 0)
    return pixel_func

def draw_rect(x1, y1, x2, y2, color):
    """Return a pixel function that draws a rectangle."""
    def pixel_func(x, y):
        if x1 <= x <= x2 and y1 <= y <= y2:
            return color
        return (0, 0, 0, 0)
    return pixel_func

def combine(*funcs):
    """Combine multiple pixel functions."""
    def combined(x, y):
        for func in reversed(funcs):
            r, g, b, a = func(x, y)
            if a > 128:
                return (r, g, b, a)
        return (0, 0, 0, 0)
    return combined

# Color definitions
GRAY = (138, 147, 132, 255)  # #8a9384
GREEN = (47, 125, 79, 255)   # #2f7d4f
WHITE = (255, 255, 255, 255)

def make_group_icon(color):
    """Create group icon (two people)."""
    size = 81
    funcs = []

    # Person 1 - head
    funcs.append(draw_circle(28, 25, 10, color))
    # Person 1 - body
    funcs.append(draw_rect(16, 40, 40, 65, color))

    # Person 2 - head
    funcs.append(draw_circle(54, 25, 10, color))
    # Person 2 - body
    funcs.append(draw_rect(42, 40, 66, 65, color))

    return combine(*funcs)

def make_checkin_icon(color):
    """Create check-in icon (calendar with checkmark)."""
    size = 81
    funcs = []

    # Calendar base (outline)
    funcs.append(draw_rect(15, 20, 65, 70, color))
    # Inner white
    funcs.append(draw_rect(18, 23, 62, 67, (255, 255, 255, 255)))

    # Checkmark
    check_funcs = []
    for t in range(100):
        # Line from (25, 45) to (38, 58)
        x = 25 + (38-25) * t / 99
        y = 45 + (58-45) * t / 99
        check_funcs.append(lambda x, y, t=t: (
            color[0], color[1], color[2],
            255 if abs(x - (25 + (38-25) * t / 99)) < 3 and abs(y - (45 + (58-45) * t / 99)) < 3 else 0
        )[3] if abs((x - (25 + (38-25) * t / 99))) < 3 else (0, 0, 0, 0))

    return combine(*funcs)

def simple_icon(pattern, color):
    """Create a simple icon based on a pattern."""
    size = 81

    def pixel_func(x, y):
        # Normalize to 0-1
        nx = x / size
        ny = y / size

        if pattern == 'group':
            # Two overlapping circles
            d1 = ((nx - 0.35)**2 + (ny - 0.35)**2)**0.5
            d2 = ((nx - 0.65)**2 + (ny - 0.35)**2)**0.5
            if d1 < 0.15 or d2 < 0.15:
                return color
            # Bodies
            if (0.2 < nx < 0.5 and 0.5 < ny < 0.8) or (0.5 < nx < 0.8 and 0.5 < ny < 0.8):
                return color
            return (0, 0, 0, 0)

        elif pattern == 'checkin':
            # Simple checkmark in a square
            if 0.2 < nx < 0.8 and 0.2 < ny < 0.8:
                # Border
                if x < 20 or x > 60 or y < 20 or y > 60:
                    return color
                # Checkmark
                cx, cy = x - 16, y - 16
                if 0 <= cx <= 40 and 0 <= cy <= 40:
                    # Simplified checkmark
                    if abs(cx - cy) < 3 and cx > 10:
                        return color
                    if abs(cx - (40 - cy)) < 3 and cx > 20:
                        return color
            return (0, 0, 0, 0)

        elif pattern == 'me':
            # Person silhouette
            d = ((nx - 0.5)**2 + (ny - 0.35)**2)**0.5
            if d < 0.15:
                return color
            if 0.3 < nx < 0.7 and 0.55 < ny < 0.8:
                return color
            return (0, 0, 0, 0)

        elif pattern == 'review':
            # Bar chart
            if 0.2 < nx < 0.4 and 0.4 < ny < 0.8:
                return color
            if 0.45 < nx < 0.65 and 0.3 < ny < 0.8:
                return color
            if 0.7 < nx < 0.9 and 0.2 < ny < 0.8:
                return color
            return (0, 0, 0, 0)

        return (0, 0, 0, 0)

    return pixel_func

def create_icon(pattern, color):
    """Create an icon with the given pattern and color."""
    pixel_func = simple_icon(pattern, color)

    def wrapper(x, y):
        nx = x / 81
        ny = y / 81

        if pattern == 'group':
            # Head 1
            d1 = ((x - 28)**2 + (y - 25)**2)**0.5
            if d1 <= 12:
                return color
            # Head 2
            d2 = ((x - 54)**2 + (y - 25)**2)**0.5
            if d2 <= 12:
                return color
            # Body 1
            if 16 <= x <= 40 and 40 <= y <= 65:
                return color
            # Body 2
            if 42 <= x <= 66 and 40 <= y <= 65:
                return color
            return (0, 0, 0, 0)

        elif pattern == 'checkin':
            # Calendar outline
            if 15 <= x <= 65 and 20 <= y <= 70:
                if x <= 18 or x >= 62 or y <= 23 or y >= 67:
                    return color
                # Checkmark
                if 25 <= x <= 55 and 35 <= y <= 55:
                    # Simplified checkmark
                    if abs((x - 25) - (y - 35) * 0.6) < 3:
                        return color
            return (0, 0, 0, 0)

        elif pattern == 'me':
            # Head
            d = ((x - 40)**2 + (y - 25)**2)**0.5
            if d <= 14:
                return color
            # Body
            if 22 <= x <= 58 and 48 <= y <= 70:
                return color
            return (0, 0, 0, 0)

        elif pattern == 'review':
            # Bars
            if 18 <= x <= 34 and 45 <= y <= 70:
                return color
            if 35 <= x <= 51 and 35 <= y <= 70:
                return color
            if 52 <= x <= 68 and 25 <= y <= 70:
                return color
            return (0, 0, 0, 0)

        return (0, 0, 0, 0)

    return wrapper

# Generate icons
icons = [
    ('tab-group', 'group', GRAY),
    ('tab-group-active', 'group', GREEN),
    ('tab-checkin', 'checkin', GRAY),
    ('tab-checkin-active', 'checkin', GREEN),
    ('tab-me', 'me', GRAY),
    ('tab-me-active', 'me', GREEN),
    ('tab-review', 'review', GRAY),
    ('tab-review-active', 'review', GREEN),
]

output_dir = 'program/images/tab-icons'
os.makedirs(output_dir, exist_ok=True)

for filename, pattern, color in icons:
    pixel_func = create_icon(pattern, color)
    png_data = create_png(81, 81, pixel_func)

    filepath = os.path.join(output_dir, f'{filename}.png')
    with open(filepath, 'wb') as f:
        f.write(png_data)
    print(f'Created: {filepath}')

print('\nDone!')
